Added QueryStringDecoder.rawPath() and rawQuery()

nedtwigg opened the pull request:
Motivation:

Before this commit, it is impossible to access the path component of the
URI before it has been decoded.  This makes it impossible to distinguish
between the following URIs:

/user/title?key=value
/user%2Ftitle?key=value

The user could already access the raw uri value, but they had to calculate
pathEndIdx themselves, even though it might already be cached inside
QueryStringDecoder.

Result:

The user can easily and efficiently access the undecoded path and query.

fenik17 commented:
The `pathEndIdx` method is the implementation details. It's not a good idea to put it outside, imho. Maybe better add two methods to getting a raw components?

```java
    public String rawPath() {
        return uri.substring(0, pathEndIdx());
    }

    public String rawQuery() {
        return uri.substring(pathEndIdx() + 1, uri.length());
    }
```

nedtwigg commented:
That was my first instinct, except that every public method on `QueryStringDecoder` right now is calculated lazily and cached, but such an implementation of `rawPath()` and `rawQuery()` would not be.

Caching is an implementation detail, but since every method `QueryStringDecoder` currently has the same performance behavior, I figured it would be less controversial to make `pathEndIdx()` public than to add new methods that have different performance characteristics than the existing methods.

I'm happy with either, I'll leave that up to you :)

normanmaurer commented:
@nedtwigg I think I would do what @fenik17 suggested.

fenik17 commented:
@nedtwigg
> I figured it would be less controversial to make pathEndIdx() public than to add new methods that have different performance characteristics than the existing methods.

Getting raw components is not a direct destination of the `QueryStringDecoder`. I think it will not be a big problem if caching issues are left to the discretion of the user.

nedtwigg commented:
Roger, I reverted my previous commit, and added `rawPath()` and `rawQuery()`, and added some tests for `QueryStringDecoderTest`.  If you'd rather I force-pushed a branch without the revert, just lemme know.

fenik17 commented:
Add tests with getting empty string by rawQuery(), please. For example:
```java
new QueryStringDecoder("path/")
new QueryStringDecoder("path?")
new QueryStringDecoder("")
```
This is for the case when `start >= uri.length()`.

fenik17 commented:
I think it's time to squash commits and update the issue/commit description.

nedtwigg commented:
Squashed with updated commit description.

normanmaurer commented:
@nedtwigg can you please sign our icla and let me know once done:

https://netty.io/s/icla

nedtwigg commented:
Signed.

normanmaurer commented:
Cherry-picked into 4.1 (dcbbae7f9006f5b995d055702dd1c9f55f0e0c9d) and 4.0 (ac8f66bb7b029f48940e8fe165a962f84e2c531d)

normanmaurer commented:
@nedtwigg thanks!
